### PR TITLE
cdc: print log to indicate memory free, and adjust finish_scan_lock method

### LINKED ISSUE #17368: cdc add more logs to help investigate memory quota related problem
once the `Drain` is dropped, release memory quota, add some logs to show the released memory amount, to help investigate problem life easier.

### PR DESCRIPTION
**Pull request by 3AceShowHand**
### What is changed and how it works?

Issue Number: Close #17368 

What's Changed:

```commit-message
* add one log to indicate the memory quota is freed when drop the `Drain`
* free the truncated scanned event memory quota.
* refactor `finish_scan_lock` method, to remove the else branch.
* row size calculation should also consider old value
* remove some outdate todo
```

### Related changes

- [ ] PR to update `pingcap/docs`/`pingcap/docs-cn`:
- [ ] Need to cherry-pick to the release branch

### Check List

Tests 

- [ ] Unit test
- [ ] Integration test
- [ ] Manual test (add detailed scripts or steps below)
- [ ] No code

Side effects

- [ ] Performance regression: Consumes more CPU
- [ ] Performance regression: Consumes more Memory
- [ ] Breaking backward compatibility

### Release note

```release-note
None
```

### CODE REVIEW

File: components/cdc/src/channel.rs [hunk 1]
```diff
@@ -431,9 +431,8 @@ impl Drop for Drain {
         });
         block_on(&mut drain);
         let takes = start.saturating_elapsed();
-        if takes >= Duration::from_millis(200) {
-            warn!("drop Drain too slow"; "takes" => ?takes);
-        }
+        info!("drop Drain finished, free memory";"takes" => ?takes,
```
Comment (hicqu) [inline components/cdc/src/channel.rs:434]: The log doesn't show which connection the `Drain` belongs to. It's better to add this information.

**Review by hicqu — APPROVED** [review]

File: components/cdc/src/endpoint.rs [hunk 2]
```diff
@@ -1023,11 +1023,21 @@ impl<T: 'static + CdcHandle<E>, E: KvEngine, S: StoreRegionMeta> Endpoint<T, E,
         locks: BTreeMap<Key, MiniLock>,
     ) {
         let region_id = region.get_id();
-        let mut deregisters = Vec::new();
-        if let Some(delegate) = self.capture_regions.get_mut(&region_id) {
-            if delegate.handle.id == observe_id {
+        match self.capture_regions.get_mut(&region_id) {
+            None => {
+                debug!("cdc region not found on region ready (finish scan locks)"; "region_id" => region.get_id());
```
Comment (overvenus) [inline components/cdc/src/endpoint.rs:1028]:
```suggestion
                debug!("cdc region not found on region ready (finish scan locks)";
                    "region_id" => region.get_id());
```

File: components/cdc/src/service.rs [hunk 3]
```diff
@@ -338,8 +338,8 @@ impl Service {
             .unwrap_or_else(|e| {
                 warn!(
                     "cdc invalid observed start key or end key version";
-                    "downstream" => ?peer, "region_id" => request.region_id,
-                    "error" => ?e,
+                    "downstream" => ?peer, "region_id" => request.region_id, "request_id" => request.region_id,
+                    "error" => ?e, "start_key" => ?request.start_key, "end_key" => ?request.end_key,
```
Comment (overvenus) [inline components/cdc/src/service.rs:342]:
```suggestion
                    "downstream" => ?peer,
                    "region_id" => request.region_id,
                    "request_id" => request.region_id,
                    "error" => ?e,
                    "start_key" => ?request.start_key,
                    "end_key" => ?request.end_key,
```

File: components/cdc/src/channel.rs [hunk 4]
```diff
@@ -207,6 +207,7 @@ pub fn channel(buffer: usize, memory_quota: Arc<MemoryQuota>) -> (Sink, Drain) {
             unbounded_receiver,
             bounded_receiver,
             memory_quota,
+            conn_id: Default::default(),
```
Comment (overvenus) [inline components/cdc/src/channel.rs:210]: If a `conn_id` is needed, try:

```rust
pub fn channel(conn_id: ConnID, buffer: usize, memory_quota: Arc<MemoryQuota>) -> (Sink, Drain)
```

File: components/cdc/src/channel.rs [hunk 5]
```diff
@@ -355,13 +356,19 @@ pub struct Drain {
...
 impl<'a> Drain {
+    pub fn set_conn_id(&'a mut self, conn_id: ConnId) {
+        self.conn_id = conn_id;
+    }
+
     pub fn drain(&'a mut self) -> impl Stream<Item = (CdcEvent, usize)> + 'a {
         let observed = (&mut self.unbounded_receiver).map(|x| (x.created, x.event, x.size));
         let scaned = (&mut self.bounded_receiver).filter_map(|x| {
             if x.truncated.load(Ordering::Acquire) {
+                self.memory_quota.free(x.size as _);
```
Comment (overvenus) [inline components/cdc/src/channel.rs:371]: Is this a bug-fix?

File: components/cdc/src/delegate.rs [hunk 6]
```diff
@@ -769,7 +769,7 @@ impl Delegate {
                     }
                     decode_default(default.1, &mut row, &mut _has_value);
                     row.old_value = old_value.finalized().unwrap_or_default();
-                    row_size = row.key.len() + row.value.len();
+                    row_size = row.key.len() + row.value.len() + row.old_value.len();
```
Comment (overvenus) [inline components/cdc/src/delegate.rs:772]: Bug-fix?

File: components/cdc/src/channel.rs (hunk 5 above)
Comment (3AceShowHand) [reply components/cdc/src/channel.rs:371]: yes I think

File: components/cdc/src/channel.rs (hunk 4 above)
Comment (3AceShowHand) [reply components/cdc/src/channel.rs:210]: The `conn_id` can only be obtained after call `Conn::new(sink, ctx.peer())`. to create the `conn`, we have to create the `sink` first...

File: components/cdc/src/service.rs [hunk 7]
```diff
@@ -338,8 +338,12 @@ impl Service {
             .unwrap_or_else(|e| {
                 warn!(
                     "cdc invalid observed start key or end key version";
-                    "downstream" => ?peer, "region_id" => request.region_id,
+                    "downstream" => ?peer,
+                    "region_id" => request.region_id,
+                    "request_id" => request.region_id,
                     "error" => ?e,
+                    "start_key" => ?request.start_key,
+                    "end_key" => ?request.end_key,
```
Comment (overvenus) [inline components/cdc/src/service.rs:346]: Please redact.

File: components/cdc/src/service.rs [hunk 8]
```diff
@@ -405,6 +409,7 @@ impl Service {
             channel(CDC_CHANNLE_CAPACITY, self.memory_quota.clone());
         let conn = Conn::new(event_sink, ctx.peer());
         let conn_id = conn.get_id();
+        event_drain.set_conn_id(conn_id);
```
Comment (overvenus) [inline components/cdc/src/service.rs:412]: How about creating conn_id separately? 

```suggestion
        let conn_id = ConnId::default();
        let (event_sink, mut event_drain) =
            channel(conn_id, CDC_CHANNLE_CAPACITY, self.memory_quota.clone());
        let conn = Conn::new(conn_id, event_sink, ctx.peer());
```

**Review by overvenus — COMMENTED** [review]
Rest LGTM

File: components/cdc/src/service.rs [hunk 9]
```diff
@@ -401,10 +405,11 @@ impl Service {
...
     ) {
         sink.enhance_batch(true);
+        let conn_id = ConnId::new();
         let (event_sink, mut event_drain) =
-            channel(CDC_CHANNLE_CAPACITY, self.memory_quota.clone());
-        let conn = Conn::new(event_sink, ctx.peer());
-        let conn_id = conn.get_id();
+            channel(conn_id, CDC_CHANNLE_CAPACITY, self.memory_quota.clone());
+        let conn = Conn::new(conn_id, event_sink, ctx.peer());
+        event_drain.set_conn_id(conn_id);
```
Comment (overvenus) [inline components/cdc/src/service.rs:412]:
```suggestion
```

File: components/cdc/src/channel.rs [hunk 10]
```diff
@@ -355,13 +356,19 @@ pub struct Drain {
...
     bounded_receiver: Receiver<ScanedEvent>,
     memory_quota: Arc<MemoryQuota>,
+    conn_id: ConnId,
 }
 
 impl<'a> Drain {
+    pub fn set_conn_id(&'a mut self, conn_id: ConnId) {
+        self.conn_id = conn_id;
+    }
+
```
Comment (overvenus) [inline components/cdc/src/channel.rs:366]:
```suggestion
```

**Review by overvenus — APPROVED** [review]

**Pull request merged**